[motion-path] Implement coord-box for ray

nmoucht opened the pull request:
#### 79256d79a14c7bf6a0cd1b8ad56b3652669c48c7
<pre>
[motion-path] Implement coord-box for ray
<a href="https://bugs.webkit.org/show_bug.cgi?id=260814">https://bugs.webkit.org/show_bug.cgi?id=260814</a>
rdar://114577300

Reviewed by Tim Nguyen.

Parse coord-box when included with ray, and use it to determine
the correct containg block box rect.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/animation/offset-path-interpolation-005-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-parsing-valid-expected.txt:
* Source/WebCore/css/CSSRayValue.cpp:
(WebCore::CSSRayValue::customCSSText const):
(WebCore::CSSRayValue::equals const):
* Source/WebCore/css/CSSRayValue.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForPathOperation):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeBasicShapeRayOrBox):
(WebCore::CSSPropertyParserHelpers::consumePathOperation):
(WebCore::CSSPropertyParserHelpers::consumeBasicShapeOrBox): Deleted.
* Source/WebCore/rendering/MotionPath.cpp:
(WebCore::offsetFromContainer):
(WebCore::MotionPath::motionPathDataForRenderer):
(WebCore::MotionPath::computePathForBox):
* Source/WebCore/rendering/PathOperation.h:
(WebCore::PathOperation::setReferenceBox):
(WebCore::PathOperation::referenceBox const):
(WebCore::PathOperation::PathOperation):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertRayPathOperation):
(WebCore::Style::BuilderConverter::convertPathOperation):

Canonical link: <a href="https://commits.webkit.org/267566@main">https://commits.webkit.org/267566@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ecc0a926695d1beea4ad781b33ba1ab39352ccce

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17004 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17327 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17794 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18787 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15910 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17194 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20560 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17466 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17205 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17551 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14731 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19602 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14793 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15425 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22140 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15778 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15592 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19919 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16192 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13728 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15352 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19716 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2092 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16025 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSRayValue.cpp:
```diff
@@ -43,7 +46,8 @@ String CSSRayValue::customCSSText() const
         hasPosition ? m_position->cssText() : ""_s,
-        ')'
+        ')',
+        stream.release()
```
```
bool hasCustomCoordinateBox = m_coordBox != CSSBoxType::BoxMissing;
```
then:
```suggestion
        hasCustomCoordinateBox ? ' ' : ""_s,
        hasCustomCoordinateBox ? m_coordBox : ""_s
```

nmoucht replied:
This is taking advantage of the `TextStream& operator<<(TextStream& ts, CSSBoxType boxType)` function. I'm not sure there is anything to convert the `CSSBoxType` to string.

nt1m replied:
You can use `nameString(toCSSValueID(m_coordBox))` or potentially just `nameString(m_coordBox)`

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/CSSRayValue.h:
```diff
@@ -42,12 +43,18 @@ class CSSRayValue final : public CSSValue {
         return adoptRef(*new CSSRayValue(WTFMove(angle), size, isContaining, WTFMove(position)));
     }
 
+    static Ref<CSSRayValue> create(Ref<CSSPrimitiveValue>&& angle, CSSValueID size, bool isContaining, RefPtr<CSSValuePair>&& position, CSSBoxType coordBox)
```
Can we avoid abbreviations here and throughout the patch? (coordinateBox)

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -706,6 +707,39 @@ inline TextAlignLast BuilderConverter::convertTextAlignLast(BuilderState& builde
+{
+    auto& rayValue = downcast<CSSRayValue>(value);
+
+    RayPathOperation::Size size = RayPathOperation::Size::ClosestCorner;
```
```suggestion
    auto size = RayPathOperation::Size::ClosestCorner;
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -7091,18 +7066,47 @@ static RefPtr<CSSRayValue> consumeRayShape(CSSParserTokenRange& range, const CSS
+    bool funcFound = false;
+    bool boxFound = false;
+    while (!range.atEnd() && !(funcFound && boxFound)) {
+        RefPtr<CSSValue> componentValue = nullptr;
```
```suggestion
        RefPtr<CSSValue> componentValue;
```

This should silence the style checker.

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/CSSRayValue.cpp:
```diff
@@ -43,7 +46,9 @@ String CSSRayValue::customCSSText() const
-        ')'
+        ')',
+        hasCustomCoordinateBox ? " "_s : ""_s,
+        hasCustomCoordinateBox ? nameString(toCSSValueID(m_coordinateBox)) : ""_s
```
```suggestion
        hasCustomCoordinateBox ? nameLiteral(toCSSValueID(m_coordinateBox)) : ""_s
```

I think this is better (feel free to change before landing) since it avoids constructing a string.

nmoucht replied:
Fixed.

nmoucht committed:
[motion-path] Implement coord-box for ray
https://bugs.webkit.org/show_bug.cgi?id=260814
rdar://114577300

Reviewed by Tim Nguyen.

Parse coord-box when included with ray, and use it to determine
the correct containg block box rect.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/animation/offset-path-interpolation-005-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-parsing-valid-expected.txt:
* Source/WebCore/css/CSSRayValue.cpp:
(WebCore::CSSRayValue::customCSSText const):
(WebCore::CSSRayValue::equals const):
* Source/WebCore/css/CSSRayValue.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForPathOperation):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeBasicShapeRayOrBox):
(WebCore::CSSPropertyParserHelpers::consumePathOperation):
(WebCore::CSSPropertyParserHelpers::consumeBasicShapeOrBox): Deleted.
* Source/WebCore/rendering/MotionPath.cpp:
(WebCore::offsetFromContainer):
(WebCore::MotionPath::motionPathDataForRenderer):
(WebCore::MotionPath::computePathForBox):
* Source/WebCore/rendering/PathOperation.h:
(WebCore::PathOperation::setReferenceBox):
(WebCore::PathOperation::referenceBox const):
(WebCore::PathOperation::PathOperation):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertRayPathOperation):
(WebCore::Style::BuilderConverter::convertPathOperation):

Canonical link: https://commits.webkit.org/267566@main
